simulator/infra/ports/module.cpp

```diff
@@ -41,9 +47,64 @@ void Module::enable_logging_impl( const std::unordered_set<std::string>& names)
         c->enable_logging_impl( names);
 }
 
+void Module::module_dumping( pt::ptree* modules) const
+{
+    pt::ptree module;
+    pt::ptree write_ports;
+    pt::ptree read_ports;
+    for ( const auto& k : topology_read_ports)
+        read_ports.put(k, "");
+    for ( const auto& k : topology_write_ports)
+        write_ports.put(k, "");
+    module.add_child( "write_ports", write_ports);
+    module.add_child( "read_ports", read_ports);
+    modules->add_child( name, module);
+    for ( const auto& c : children)
+        c->module_dumping( modules);
+}
+
+void Module::modulemap_dumping( pt::ptree* modulemap) const
+{
+    pt::ptree c_modulemap;
+    for ( const auto& c : children)
+        c->modulemap_dumping( &c_modulemap);
+    modulemap->add_child( name, c_modulemap);
+}
+
 void Root::enable_logging( const std::string& values)
 {
     std::unordered_set<std::string> tokens;
     boost::split( tokens, values, boost::is_any_of( std::string(",")));
     enable_logging_impl( tokens);
 }
+
+void Root::portmap_dumping( pt::ptree* pt_portmap) const
+{
+    portmap->dump( pt_portmap);
+}
+
+void Root::topology_dumping_impl( pt::ptree* topology) const
+{
+    pt::ptree modules;
+    pt::ptree pt_portmap;
+    pt::ptree modulemap;
+
+    module_dumping( &modules);
+    portmap_dumping( &pt_portmap);
+    modulemap_dumping( &modulemap);
```

Comment:
Can these methods return pt::ptree instead of passing as arguments? That would reduce code to the 3 simple lines.

Reply:
I changed only `portmap_dumping`. Others can't save the dump format because boost ptree doesn't allow me to insert the recursive subtree correctly.